torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -0,0 +1,295 @@
+
+// Run time check
+#if defined(MPIX_CUDA_AWARE_SUPPORT)
+  if (MPIX_Query_cuda_support() == 1) {
+    return true;
+  } else {
+    return false;
+  }
+#else /* !defined(MPIX_CUDA_AWARE_SUPPORT) */
+  return false;
+#endif /* MPIX_CUDA_AWARE_SUPPORT */
+}
+
+} // namespace
+
+
+// ProcessGroupMPI::WorkMPI
+ProcessGroupMPI::WorkMPI::WorkMPI() : completed_(false) {}
+
+ProcessGroupMPI::WorkMPI::~WorkMPI() {}
+
+bool ProcessGroupMPI::WorkMPI::isCompleted() const {
+  return completed_;
+}
+
+bool ProcessGroupMPI::WorkMPI::isSuccess() const {
+  return !ex_;
+}
+
+bool ProcessGroupMPI::WorkMPI::wait() {
+  std::unique_lock<std::mutex> lock(workMutex_);
+  while (!completed_) {
+    workCV_.wait(lock);
+  }
+  return isSuccess();
+}
+
+void ProcessGroupMPI::WorkMPI::finish() {
+  {
+    std::unique_lock<std::mutex> lock(workMutex_);
+    completed_ = true;
+  }
+  workCV_.notify_all();
+}
+
+void ProcessGroupMPI::WorkMPI::finishWithException(
+    const std::exception& ex) {
+  {
+    std::unique_lock<std::mutex> lock(workMutex_);
+    completed_ = true;
+    ex_ = std::unique_ptr<std::exception>(new std::exception(ex));
+  }
+  workCV_.notify_all();
+}
+
+const std::exception& ProcessGroupMPI::WorkMPI::exception() const {
+  return *ex_;
+}
+
+// ProcessGroupMPI
+
+// Static global states
+int ProcessGroupMPI::numProcessGroups_ = 0;
+int ProcessGroupMPI::mpiThreadSupport_ = 0;
+bool ProcessGroupMPI::mpiInitialized_ = false;
+std::mutex ProcessGroupMPI::pgGlobalMutex_;
+
+ProcessGroupMPI::ProcessGroupMPI()
+: ProcessGroup(-1, 0)
+, stop_(false)
+{
+  initMPI();
+  workerThread_ = std::thread(&ProcessGroupMPI::runLoop, this);
+}
+
+ProcessGroupMPI::~ProcessGroupMPI() {
+  destroy();
+}
+
+void ProcessGroupMPI::initMPI() {
+  std::unique_lock<std::mutex> globalLock(pgGlobalMutex_);
+  // We only want to initialize once
+  try {
+    if (!mpiInitialized_) {
+      MPI_CHECK(MPI_Init_thread(nullptr,
+                                nullptr,
+                                MPI_THREAD_MULTIPLE,
+                                &mpiThreadSupport_));
+      if (mpiThreadSupport_ < MPI_THREAD_SERIALIZED) {
+        throw std::runtime_error("Used MPI implementation doesn't have the "
+                                 "minimum level of threading support: "
+                                 "MPI_THREAD_SERIALIZED. This is required by "
+                                 "c10d package");
+      }
+      if (mpiThreadSupport_ != MPI_THREAD_MULTIPLE) {
+        std::cerr << "WARNING: Used MPI implementation doesn't support "
+                  << "multithreading, so distributed cannot support more "
+                  << "than one process group work."
+                  << std::endl;
+      }
+      mpiInitialized_ = true;
+    }
+
+    // Update the world size and rank
+    MPI_CHECK(MPI_Comm_size(MPI_COMM_WORLD, &size_));
+    MPI_CHECK(MPI_Comm_rank(MPI_COMM_WORLD, &rank_));
+
+  } catch (...) {
+    MPI_CHECK(MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE));
```

Comment:
We should not abort the whole program unconditionally using `MPI_Abort`, because it will suppress any any errors you raised from inside this function (including the one about thread level support).